content/blog/2022/2022-09-25-node-urlShortener.md

```diff
@@ -0,0 +1,494 @@
+---
+title: "Building a Url Shortener With Node.Js"
+categories: ["Node"]
+date: 2022-09-25 00:00:00 +1100
+authors: ["ajibade"]
+description: "Building A MERN stack Url Shortener Application"
+image: images/stock/0043-calendar-1200x628-branded.jpg
+url: node-url-shortener
+---
+
+
+The best URL shortening service depends on your needs and what you want it to do. There are a lot of simple, fast, and free URL shortening services. Some provide detailed information about who is clicking your links, the number of times they are clicked and even options to add a [call to action](https://en.wikipedia.org/wiki/Call_to_action_(marketing)) to links. URL shorteners like [bitly](https://bitly.com/) and [cutt.ly](https://cutt.ly/) are popular examples.
+
+This tutorial will walk you through how to develop a URL shortener application using the `MERN stack`. Developing `REST APIs` with `Node.js` and `MongoDB` and then integrating them into `React.js` frontend applications. This is a common coding interview question and an excellent practice project for new `Node.js` and `React` developers.
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/nodejs/url-shortener" %}}
+
+## Urlshortner Behind the Scene
+A URL shortener service allows users to enter long URLs and generate shorter, unique URLs in return. When the user lookup generated short URLs, they are redirected to the original long URL content.
+
+First, a URL shortener service selects a short domain name as a placeholder. Such as `goo.gl` `tinyurl.com` or `bit.ly`  Then a  unique shortcode or id is generated. For example, the URL https://reflectoring.io/schedule-cron-job-in-node/ can be shortened to https://lnkd.in/dFZXFq2T.
+
+These generated shortcodes are stored alongside the domain name.
+
+And then mapped to the long URL.
+
+When we visit the newly generated short URL. A database lookup is performed using the shortcode as a key, and if it is found, we are redirected to the long URL's content.
+
+URL shortener service simply uses a URL redirect. However, scaling, monitoring, URL filtering, spam prevention, URL verification, and other required features can result in more advanced logics.
+
+## Setting up `Node.Js` Application
```

Comment:
Before we jump into the Node.js backend, it would be good if we had a section "URL Shortener Architecture" that shows a diagram of the components (backend, database, frontend, ...) and explains each of the components in a sentence or two. Otherwise, the reader doesn't know what we're building.